docs/src/adr/014adr-precise-records.md

```diff
@@ -0,0 +1,668 @@
+              ...
+```           
+
+Occurences in [tlaplus-examples](https://github.com/tlaplus/Examples):
+
+ - [Paxos](https://github.com/tlaplus/Examples/blob/master/specifications/Paxos/Paxos.tla)
+ - [802.16](https://github.com/tlaplus/Examples/tree/master/specifications/802.16)
+ - [byihive](https://github.com/byisystems/byihive/blob/master/specifications/tla%2B/VoucherIssue.tla)
+ - [CASPaxos](https://github.com/tbg/caspaxos-tla/blob/master/CASPaxos.tla)
+ - [cbc_max](https://github.com/tlaplus/Examples/blob/master/specifications/cbc_max/cbc_max.tla)
+ - [EgalitarianPaxos](https://github.com/efficient/epaxos/blob/master/tla%2B/EgalitarianPaxos.tla)
+ - [FPaxos](https://github.com/fpaxos/fpaxos-tlaplus/blob/master/FPaxos.tla)
+ - [raft](https://github.com/ongardie/raft.tla/blob/master/raft.tla)
+ - [SnapshotIsolation](https://github.com/sanjosh/tlaplus/blob/master/amazon/serializableSnapshotIsolation.tla)
+ - [PaxosCommit](https://github.com/tlaplus/Examples/blob/master/specifications/transaction_commit/PaxosCommit.tla)
+
+
+More advanced code patterns over variants can be found in [Raft][]:
+
+```tla
+Receive(m) ==
+    LET i == m.mdest
+        j == m.msource
+    IN \* Any RPC with a newer term causes the recipient to advance
+       \* its term first. Responses with stale terms are ignored.
+       \/ UpdateTerm(i, j, m)
+       \/ /\ m.mtype = RequestVoteRequest
+          /\ HandleRequestVoteRequest(i, j, m)
+       \/ /\ m.mtype = RequestVoteResponse
+          /\ \/ DropStaleResponse(i, j, m)
+             \/ HandleRequestVoteResponse(i, j, m)
+       \/ /\ m.mtype = AppendEntriesRequest
+          /\ HandleAppendEntriesRequest(i, j, m)
+       \/ /\ m.mtype = AppendEntriesResponse
+          /\ \/ DropStaleResponse(i, j, m)
+             \/ HandleAppendEntriesResponse(i, j, m)
+    ...
+Next ==
+    ...
+           \/ \E m \in DOMAIN messages : Receive(m)
+           \/ \E m \in DOMAIN messages : DuplicateMessage(m)
+           \/ \E m \in DOMAIN messages : DropMessage(m)
+    ...
+```
+
+### 3. Other issues
+
+As can be seen from the Paxos example, we should take care of sets of
+records that are used as sets of variants:
+
+```tla
+Message ==      [type : {"1a"}, bal : Ballot]
+           \cup [type : {"1b"}, acc : Acceptor, bal : Ballot, 
+                 mbal : Ballot \cup {-1}, mval : Value \cup {None}]
+           \cup [type : {"2a"}, bal : Ballot, val : Value]
+           \cup [type : {"2b"}, acc : Acceptor, bal : Ballot, val : Value]
+```           
+
+## 4. Options
+
+There are several solutions to the issue of precise record typing.
+
+### 4.1. Support only plain records
+
+In this case, we would only allow to mix records that have exactly the same
+shape. As a result, when we use variants, we would have to add extra
+fields to all of them. This solution is not very different from the current
+implementation of record type checking, though it would allow us to quickly
+detect spelling errors.
```

Comment:
The moment you use `G(dinner)`, you should get a type error. If `dinner` appears nowhere, then `G` is just a polymorphic operator.

Reply:
Oh man. Then I am really confused. I thought this was exactly the thing that the row types were for. In the cited paper, they give this example:

> The selection operation (r .l ) selects the value of a label
> l from a record r . For example, we can define a function distance
> that calculates the distance of a point to the origin:
> distance p = sqrt ((p.x ∗ p.x ) + (p.y ∗ p.y))
> In contrast to many programming languages, the distance function
> works for any record that contains an x and y field of a suitable
> numeric type.

My focus here is on this point:

> the distance function works for any record that contains an x and y field of a suitable numeric type.

So I was expecting that `G` and `F` would the same, namely that `F` would work on any record that contains a `food` of a suitable `Str` type, and that `G`, since it's just doing a record extension, would work to add the field `fool` to any record that (simply because of the type annotation) has a `foo : Str`. But I think you've noted elsewhere that you don't plan to support record extension? Yeah, I am quite confused.

Reply:
I talked with Shon on slack about `EXCEPT` behavior because it was not clear to us whether it could be used for extending records. Just to register here: it's not, at least not on TLC. Trying to extend a record gives a warning:
```
Warning: The EXCEPT was applied to non-existing fields of the value
```
and the operation is ignored.

Reply:
`EXCEPT` won't change the function domain, see https://www.microsoft.com/en-us/research/uploads/prod/2018/05/book-02-08-08.pdf (p302, 16.7)

Reply:
Right. In TLA+, records are functions, and functions are not allowed to extend their domains. So, yes, records are not allowed to extend their domains.

Reply:
In my understanding, row types allow to infer types of records whose type is not fully known inside an operator (or a function in a functional language). So we would have to produce a polymorphic record, whose type can be instantiated with a record that has more fields. Whether the language itself supports record extension or not, is less important, though record extension probably makes things more complicated.

Reply:
Ah yeah, I see. 

Having that narrower scope helps me understand the stakes and aim here better. I wonder if you that framing and narrow focus could be added to the ADR? I'm not sure if it's just me, but I think many of my confusions here would have been avoided if the setup noted these significant deviations and limits (compared against the paper) up front. I'll suggest that change below.

Thanks, all (especially @gabrielamafra), for helping me walk through this bit!